File: FlattenBinarytreeLinkedList.java
// Time Complexity : O(n); n = #nodes
// Space Complexity : O(n); recursion stack
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

public class FlattenBinarytreeLinkedList {
    public void flatten(TreeNode root) {
        helper(root);
    }
    private TreeNode helper(TreeNode root) {
        //edge case
        if (root == null) return null;

        //base case
        if (root.left == null && root.right == null) return root;

        //logic

        TreeNode left = helper(root.left);
        TreeNode right = helper(root.right);

        if (left != null) {
            left.right = root.right;
            root.right = root.left;
            root.left = null;
        }

        if (right != null) return right;
        else return left;
    }
}

File: FlattenBinaryTree_LinkedList.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    """without using stack and recursion--no extra space used---Top Down approach
    Time complexity-O(n) as going through all elements of tree
    Space complexity-O(1)"""
    def __init(self):
        self.temp=None
    def flatten(self, root: TreeNode) -> None:
        if not root or (not root.left and not root.right):
            return
        while root:
            self.temp=root.right
            if root.left:
                root.right=root.left
                root.left=None
                if root.right:
                    updateroot=root.right
                    while updateroot.right!=None:
                        updateroot=updateroot.right
                updateroot.right=self.temp
            root=root.right
    # def flatten(self, root: TreeNode) -> None:
        """Using recursive stack
        Time complexity-O(n)
        Space complexity-O(h)"""
        # if not root:
        #     return
        # self.flatten(root.left)
        # temp=root.right
        # if root.left:
        #     root.right=root.left
        #     root.left=None
        #     temp1=root.right
        #     if temp1:
        #         while temp1.right:
        #             temp1=temp1.right
        #         temp1.right=temp
        # self.flatten(root.right)
        
        